### PR TITLE
wasi-threads: basic simplistic implementation of wasi-threads proposal

### PR DESCRIPTION
**Pull request by loganek**
wasi-threads proposal: https://github.com/WebAssembly/wasi-threads

This is very very basic implementation, probably doesn't cover a lot of corner cases. I open the PR here to get early feedback on the direction, missing bits and suggestions.

Also, I think it'd be good to have a separate branch for that (dev/wasi-threads?) so if we can have it setup that'd be great.

### CODE REVIEW

**Comment by wenyongh** [comment]
> wasi-threads proposal: https://github.com/WebAssembly/wasi-threads
> 
> This is very very basic implementation, probably doesn't cover a lot of corner cases. I open the PR here to get early feedback on the direction, missing bits and suggestions.
> 
> Also, I think it'd be good to have a separate branch for that (dev/wasi-threads?) so if we can have it setup that'd be great.

@loganek Thanks for uploading the PR! I have created branch `dev/wasi_threads` and changed the PR's destination branch to it.

File: samples/wasi-threads/wasm-apps/main.c [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        }                                                  \
+    } while (0)
+
+pthread_cond_t cond;
+pthread_mutex_t mp;
+
+int global_value = 0;
+
+void
+sleep(int64_t time_ns)
```
Comment (yamt) [inline samples/wasi-threads/wasm-apps/main.c:20]: please avoid overriding unistd.h sleep(3). my suggestion is sleep_ns.

File: samples/wasi-threads/wasm-apps/main.c [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        if (ret != 0) {                                    \
+            printf("Failed to call %s: %d\n", #call, ret); \
+            return -1;                                     \
+        }                                                  \
+    } while (0)
+
+pthread_cond_t cond;
+pthread_mutex_t mp;
+
+int global_value = 0;
```
Comment (yamt) [inline samples/wasi-threads/wasm-apps/main.c:17]: volatile?

File: core/iwasm/libraries/lib-wasi-threads/lib_wasi_threads_wrapper.c [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    info = wasm_runtime_malloc(sizeof(ThreadInfo));
+    if (!info)
+        goto fail;
+
+
+    info->start_arg = start_arg;
+    info->start_func = start_func;
+    info->exec_env = new_exec_env;
+
+    ret = pthread_create(&th, NULL, thread_start, info);
```
Comment (yamt) [inline core/iwasm/libraries/lib-wasi-threads/lib_wasi_threads_wrapper.c:135]: i guess you should use os_thread_create.

**Review by yamt — COMMENTED** [review]
lgtm

File: samples/wasi-threads/wasm-apps/main.c (hunk 2 above)
Comment (yamt) [reply samples/wasi-threads/wasm-apps/main.c:17]: answering myself, volatile isn't necessary here as surrounded by mutex calls.

**Comment by loganek** [comment]
Resolving this PR in favor of the https://github.com/bytecodealliance/wasm-micro-runtime/pull/1783 PR (and further PRs).

**Pull request closed without merging**